test/UsedTxOuts.purs

```diff
@@ -0,0 +1,128 @@
+  , unlockTransactionInputs
+  , unlockTxOutRefs
+  )
+
+buildSampleTransaction
+  :: { tx :: Transaction
+     , usedTxOutRefs :: Array { transaction_id :: TransactionHash, index :: UInt }
+     , unusedTxOutRefs :: Array { transaction_id :: TransactionHash, index :: UInt }
+     }
+buildSampleTransaction =
+  let
+    usedTxOutRefs =
+      [ mkTxInput { txId: "5d677265fa5bb21ce6d8c7502aca70b9316d10e958611f3c6b758f65ad959996", ix: 0 }
+      , mkTxInput { txId: "5d677265fa5bb21ce6d8c7502aca70b9316d10e958611f3c6b758f65ad959996", ix: 1 }
+      , mkTxInput { txId: "92c4f22371bd453aec9fe19ccebfbc88211ae854b5eab424bcd4c26d4c9754cc", ix: 0 }
+      , mkTxInput { txId: "92c4f22371bd453aec9fe19ccebfbc88211ae854b5eab424bcd4c26d4c9754cc", ix: 1 }
+      , mkTxInput { txId: "92c4f22371bd453aec9fe19ccebfbc88211ae854b5eab424bcd4c26d4c9754cc", ix: 2 }
+      , mkTxInput { txId: "c9e9c15d4f16a7948d3736c93aa79034621d51dccc4df5d31c7d34aa0b3240cc", ix: 1 }
+      ]
+    unusedTxOutRefs =
+      [ mkTxInput { txId: "a69fdfbdab33eeb05143bf3b05591679fb104157fa68fe3f4268631aedf5e607", ix: 0 }
+      , mkTxInput { txId: "a69fdfbdab33eeb05143bf3b05591679fb104157fa68fe3f4268631aedf5e607", ix: 1 }
+      , mkTxInput { txId: "a2cb07ae236c4d8ee549d1bf3ad3818d691a5a7c53c1065276bfbb0ac332ccd7", ix: 0 }
+      ]
+  in
+    { tx: mkSampleTx txFixture1 (_ { inputs = usedTxOutRefs })
+    , usedTxOutRefs: unwrap <$> usedTxOutRefs
+    , unusedTxOutRefs: unwrap <$> unusedTxOutRefs
+    }
+
+suite :: TestPlanM Unit
+suite =
+  group "UsedTxOuts api tests" do
+
+    let
+      { tx, usedTxOutRefs, unusedTxOutRefs } = buildSampleTransaction
+      anyTxOutsLocked txos = any identity <$> traverse isTxOutRefUsed txos
+      allTxOutsLocked txos = all identity <$> traverse isTxOutRefUsed txos
+
+    test "UsedTxOuts cache properly locks and unlocks tx txouts" $ do
+      newUsedTxOuts >>= runReaderT
+        ( do
+            -- starts empty
+            log "TxOuts should start unlocked"
+            anyTxOutsLocked (usedTxOutRefs <> unusedTxOutRefs) `shouldReturn` false
+
+            -- lock
+            lockTransactionInputs tx
+            log "All usedTxOuts should be now locked"
+            allTxOutsLocked usedTxOutRefs `shouldReturn` true
+            log "None of the unused should be locked"
+            anyTxOutsLocked unusedTxOutRefs `shouldReturn` false
+
+            -- unlock
+            unlockTransactionInputs tx
+            log "All TxOuts should be now unlocked"
+            anyTxOutsLocked (usedTxOutRefs <> unusedTxOutRefs) `shouldReturn` false
+        )
+
+    test "UsedTxOuts cache properly locks and unlock selected txouts" $ do
+      newUsedTxOuts >>= runReaderT
+        ( do
+            -- starts empty
+            log "TxOuts should start unlocked"
+            anyTxOutsLocked (usedTxOutRefs <> unusedTxOutRefs) `shouldReturn` false
+
+            -- lock
+            lockTransactionInputs tx
+            log "All usedTxOuts should be now locked"
```

Comment:
Can we avoid logging to console in tests? It becomes annoying when there's a large suite. Would be ideal if there was a solution that outputs collected logs only in case of a test failure. However, since we already do console logging in a couple of places, feel free not to change anything.

Reply:
because it is being run in a readert, I am not able to use test instead. Should we have an issue for these smart logging capabilities?

Reply:
I think It's possible to deal with `ReaderT` in definitions of `anyTxOutsLocked` and `allTxOutsLocked` and pass `UsedTxOuts` to them as an extra argument.

Reply:
Do you suggest providing both pure and monadask based API?

Reply:
I'd be against passing newtyped ref for an update like this: `f :: UsedTxOuts -> Effect Unit`

Reply:
> Do you suggest providing both pure and monadask based API?

No, I'm talking about changing just the tests. See line 63 of this file. `ReaderT` can be added there, and both functions can accept a pure value of `UsedTxOuts`.

NB: there's an instance of `MonadAsk` for `Function`, so we get both pure and monadic interface all in one.

Reply:
Ad NB: At first sight but then we have the MonadEffect constraint -.-'

Reply:
Also (TIL) moving ReaderT to encompass calls will not work - the changes get reverted when exiting.